### PR TITLE
feat(universal-router-sdk): v3 to v4 migrator

### PR DESCRIPTION
**Pull request by marktoda**
- feat(universal-router-sdk): add UR position management commands
- feat(ur-sdk): add v3 to v4 migration
- feat: add todos

## PR Scope

Please title your PR according to the following types and scopes following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/):

- `fix(SDK name):` will trigger a patch version
- `chore(<type>):` will not trigger any release and should be used for internal repo changes
- `<type>(public):` will trigger a patch version for non-code changes (e.g. README changes)
- `feat(SDK name):` will trigger a minor version
- `feat(breaking):` will trigger a major version for a breaking change

## Description

_[Summary of the change, motivation, and context]_

## How Has This Been Tested?

_[e.g. Manually, E2E tests, unit tests, Storybook]_

## Are there any breaking changes?

_[e.g. Type definitions, API definitions]_

If there are breaking changes, please ensure you bump the major version Bump the major version (by using the title `feat(breaking): ...`), post a notice in #eng-sdks, and explicitly notify all Uniswap Labs consumers of the SDK.

## (Optional) Feedback Focus

_[Specific parts of this PR you'd like feedback on, or that reviewers should pay closer attention to]_

## (Optional) Follow Ups

_[Things that weren't addressed in this PR, ways you plan to build on this work, or other ways this work could be extended]_

### CODE REVIEW

**Comment by marktoda** [comment]
need https://github.com/Uniswap/sdks/issues/113

File: sdks/universal-router-sdk/src/swapRouter.ts [hunk 1]
```diff
@@ -43,6 +64,45 @@ export abstract class SwapRouter {
     })
   }
 
+  /**
+   * Builds the call parameters for a migration from a V3 position to a V4 position.
+   * Some requirements of the parameters:
+   *   - v3RemoveLiquidityOptions.collectOptions.recipient must equal v4PositionManager
+   *   - v3RemoveLiquidityOptions.liquidityPercentage must be 100%
+   *   - input pool and output pool must have the same tokens
+   *   - V3 NFT must be approved, or valid inputV3NFTPermit must be provided with UR as spender
```
Comment (marktoda) [inline sdks/universal-router-sdk/src/swapRouter.ts:73]: was considering not taking `V3RemoveOptions` and `V4RemoveOptions` and rather taking minimal deduplicated params and handling these automatically.. decided for this approach to give more configurability to options for each stage

File: sdks/universal-router-sdk/src/swapRouter.ts [hunk 2]
```diff
@@ -43,6 +63,45 @@ export abstract class SwapRouter {
...
+   */
+  public static migrateV3ToV4CallParameters(options: MigrateV3ToV4Options): MethodParameters {
+    const token0 = options.inputPosition.pool.token0
+    const token1 = options.inputPosition.pool.token1
+    invariant(token0 === options.outputPosition.pool.token0, 'TOKEN0_MISMATCH')
+    invariant(token1 === options.outputPosition.pool.token1, 'TOKEN1_MISMATCH')
+    invariant(options.v3RemoveLiquidityOptions.liquidityPercentage.equalTo(new Percent(100)), 'FULL_REMOVAL_REQUIRED')
```
Comment (hensha256) [inline sdks/universal-router-sdk/src/swapRouter.ts:79]: is your plan to make invariants to check the spender is UR and the recipient is v4positionmanager once we have those addresses?

File: sdks/universal-router-sdk/src/swapRouter.ts [hunk 3]
```diff
@@ -43,6 +63,45 @@ export abstract class SwapRouter {
...
 
+  /**
+   * Builds the call parameters for a migration from a V3 position to a V4 position.
+   * Some requirements of the parameters:
+   *   - v3RemoveLiquidityOptions.collectOptions.recipient must equal v4PositionManager
+   *   - v3RemoveLiquidityOptions.liquidityPercentage must be 100%
+   *   - input pool and output pool must have the same tokens
+   *   - V3 NFT must be approved, or valid inputV3NFTPermit must be provided with UR as spender
+   */
+  public static migrateV3ToV4CallParameters(options: MigrateV3ToV4Options): MethodParameters {
```
Comment (hensha256) [inline sdks/universal-router-sdk/src/swapRouter.ts:74]: shoudl it check that the `migrate` param added to v4-sdk is true?

Comment (snreynolds) [reply sdks/universal-router-sdk/src/swapRouter.ts:72]: it should imo +

File: sdks/universal-router-sdk/src/swapRouter.ts (hunk 2 above)
Comment (snreynolds) [reply sdks/universal-router-sdk/src/swapRouter.ts:77]: can we add this now? we should be checking like

spender == UR_ADDRESS[chainId] && recipient == UR_ADDRESS[chainId] and mock the chainId in the tests

File: sdks/universal-router-sdk/src/swapRouter.ts [hunk 4]
```diff
@@ -43,6 +61,45 @@ export abstract class SwapRouter {
...
+    const token0 = options.inputPosition.pool.token0
+    const token1 = options.inputPosition.pool.token1
+    invariant(token0 === options.outputPosition.pool.token0, 'TOKEN0_MISMATCH')
+    invariant(token1 === options.outputPosition.pool.token1, 'TOKEN1_MISMATCH')
+    invariant(options.v3RemoveLiquidityOptions.liquidityPercentage.equalTo(new Percent(100)), 'FULL_REMOVAL_REQUIRED')
+
+    const planner = new RoutePlanner()
+
+    if (options.inputV3NFTPermit) {
+      // note: permit spender should be UR
```
Comment (snreynolds) [inline sdks/universal-router-sdk/src/swapRouter.ts:82]: yeah we should be checking the spender here in the invariants above

**Review by snreynolds — COMMENTED** [review]
cool looks good! we should definitely add more tests
failure cases:
- pools don't have the same tokens
- spender not router/receiver not posm

& forge tests:
- checking that the calldata we produce actually executes the txn we expect

File: sdks/universal-router-sdk/src/swapRouter.ts (hunk 4 above)
Comment (dianakocsis) [reply sdks/universal-router-sdk/src/swapRouter.ts:82]: how can i get the UR address to test?

File: sdks/universal-router-sdk/src/swapRouter.ts [hunk 5]
```diff
@@ -43,6 +67,64 @@ export abstract class SwapRouter {
...
+
+    const planner = new RoutePlanner()
+
+    if (options.inputV3NFTPermit) {
+      // permit spender should be UR
+      const universalRouterAddress = UNIVERSAL_ROUTER_ADDRESS(
+        UniversalRouterVersion.V2_0,
+        options.inputPosition.pool.chainId as SupportedChainsType
+      )
+      invariant(universalRouterAddress == options.inputV3NFTPermit.spender, 'INVALID_SPENDER')
```
Comment (snreynolds) [inline sdks/universal-router-sdk/src/swapRouter.ts:107]: sick!

File: sdks/universal-router-sdk/src/swapRouter.ts [hunk 6]
```diff
@@ -2,18 +2,42 @@ import invariant from 'tiny-invariant'
...
   sender?: string // address
   deadline?: BigNumberish
 }
 
+export interface MigrateV3ToV4Options {
+  inputPosition: V3Position
+  outputPosition: V4Position
+  v3RemoveLiquidityOptions: V3RemoveLiquidityOptions
+  v4AddLiquidityOptions: V4AddLiquidityOptions
+  inputV3NFTPermit?: V3PositionPermit
```
Comment (snreynolds) [inline sdks/universal-router-sdk/src/swapRouter.ts:34]: ah sorry I didn't catch this earlier but I think that this is an unneeded parameter. the v3RemoveLiquidityOptions defined already an optional `permit` field

File: sdks/universal-router-sdk/test/forge/interop.json [hunk 7]
```diff
@@ -151,124 +163,12 @@
...
-    "value": "954092494227022159"
-  },
-  "_UNISWAP_V3_1_ETH_FOR_USDC_WITH_FEE": {
```
Comment (snreynolds) [inline sdks/universal-router-sdk/test/forge/interop.json:218]: why is the calldata for all non migration swaps changing?

Comment (snreynolds) [reply sdks/universal-router-sdk/test/forge/interop.json:218]: i think its just a weird diff..

**Pull request merged**